tutorial/lesson_19_staging_func_or_image_param.cpp

```diff
@@ -0,0 +1,289 @@
+// Halide tutorial lesson 19: Staging Func or ImageParam
+
+// This lesson demonstrates how to use Func::in() and ImageParam::in() to
+// schedule a Func differently in different places and to stage loads from
+// a Func or an ImageParam.
+
+// On linux, you can compile and run it like so:
+// g++ lesson_19*.cpp -g -I ../include -L ../bin -lHalide -lpthread -ldl -o lesson_19 -std=c++11
+// LD_LIBRARY_PATH=../bin ./lesson_19
+
+// On os x:
+// g++ lesson_19*.cpp -g -I ../include -L ../bin -lHalide -o lesson_19 -std=c++11
+// DYLD_LIBRARY_PATH=../bin ./lesson_19
+
+// If you have the entire Halide source tree, you can also build it by
+// running:
+//    make tutorial_lesson_19_staging_func_or_image_param
+// in a shell with the current directory at the top of the halide
+// source tree.
+
+// The only Halide header file you need is Halide.h. It includes all of Halide.
+#include "Halide.h"
+
+// We'll also include stdio for printf.
+#include <stdio.h>
+
+using namespace Halide;
+
+int main(int argc, char **argv) {
+    // First we'll declare some Vars to use below.
+    Var x("x"), y("y");
+
+    // This lesson will be about "wrapping" a Func or an ImageParam using the
+    // Func::in() and ImageParam::in() directives, which may be useful for
+    // variety of use cases.
+
+    {
+        // Say we have the following pipeline:
+        Func g("g"), f1("f1"), f2("f2");
+        g(x, y) = x + y;
+        f1(x, y) = x * y * g(x, y);
+        f2(x, y) = x - y + g(x, y);
+        // and we want to schedule 'g' differently depending on whether it is
+        // used by 'f1' or 'f2': in 'f1', we want to vectorize 'g' across the
+        // x dimension, while in 'f2', we want to parallelize it. It is possible
+        // to do this in Halide using the in() directive.
+
+        // g.in(f1) returns a Func that wraps all calls of 'g' inside 'f1',
```

Comment:
Is it clearer to say "…a Func *handle* that wraps…" to clarify that this is primarily a handle for scheduling, and not about creating new algorithm logic?